include/seqan3/io/alignment_file/input.hpp

```diff
@@ -0,0 +1,1142 @@
+    {
+        assert(ref_ids.size() == ref_sequences.size());
+
+        using mytype = header_type;
+        header_ptr = std::unique_ptr<mytype>{std::make_unique<mytype>(ref_ids)};
+
+        reference_sequences_ptr = &ref_sequences;
+
+        // initialise reference map and ref_dict if ref_ids are non-empty
+        for (size_t idx = 0; idx < ref_ids.size(); ++idx)
+        {
+            header_ptr->ref_id_info.emplace_back(std::make_pair(std::ranges::size(ref_sequences[idx]), std::string{}));
+            header_ptr->ref_dict[(header_ptr->ref_ids())[idx]] = idx;
+        }
+    }
+    //!\}
+
+    //!\brief Tell the format to move to the next record and update the buffer.
+    void read_next_record()
+    {
+        // clear the record
+        record_buffer.clear();
+        detail::get_or_ignore<field::HEADER_PTR>(record_buffer) = header_ptr.get();
+
+        // at end if we could not read further
+        if (std::istreambuf_iterator<stream_char_type>{*secondary_stream} ==
+            std::istreambuf_iterator<stream_char_type>{})
+        {
+            at_end = true;
+            return;
+        }
+
+        auto call_read_func = [this] (auto & ref_seq_info)
+        {
+            std::visit([&] (AlignmentFileInputFormat & f)
+            {
+                f.read(*secondary_stream,
+                       options,
+                       ref_seq_info,
+                       *header_ptr,
+                       detail::get_or_ignore<field::SEQ>(record_buffer),
+                       detail::get_or_ignore<field::QUAL>(record_buffer),
+                       detail::get_or_ignore<field::ID>(record_buffer),
+                       detail::get_or_ignore<field::OFFSET>(record_buffer),
+                       detail::get_or_ignore<field::REF_SEQ>(record_buffer),
+                       detail::get_or_ignore<field::REF_ID>(record_buffer),
+                       detail::get_or_ignore<field::REF_OFFSET>(record_buffer),
+                       detail::get_or_ignore<field::ALIGNMENT>(record_buffer),
+                       detail::get_or_ignore<field::FLAG>(record_buffer),
+                       detail::get_or_ignore<field::MAPQ>(record_buffer),
+                       detail::get_or_ignore<field::MATE>(record_buffer),
+                       detail::get_or_ignore<field::TAGS>(record_buffer),
+                       detail::get_or_ignore<field::EVALUE>(record_buffer),
+                       detail::get_or_ignore<field::BIT_SCORE>(record_buffer));
+
+            }, format);
+        };
+
+        assert(!format.valueless_by_exception());
+
+        if constexpr (!std::Same<typename traits_type::ref_sequences, ref_info_not_given>)
+            call_read_func(*reference_sequences_ptr);
+        else
+            call_read_func(std::ignore);
+    }
+
+    //!\brief Befriend iterator so it can access the buffers.
+    friend iterator;
+};
+
+/*!\name Type deduction guides
```

Comment:
Just a heads up: doxygen 1.8.15 wants you to tell us what each type deduction does